[JENKINS-33364] Core documentation: Freestyle project config

orrc opened the pull request:
Ticket: https://issues.jenkins-ci.org/browse/JENKINS-33364
This PR covers: the freestyle project config page

For the "styleguide" and more details, see the first PR in this series: https://github.com/jenkinsci/jenkins/pull/2101

orrc committed:
Expand 'discard old builds' documentation.

oleg-nenashev commented on core/src/main/resources/jenkins/model/BuildDiscarderProperty/help.html:
```diff
@@ -1,19 +1,51 @@
+  <p>
+  <!-- TODO: Most/all of the following should be dynamic, showing info from the
+       `BuildDiscarder` implementations that are currently loaded -->
+  Jenkins offers two options for determining when builds should be discarded:
```
This is an extension point, needs to be mentioned. Currently it smells like 2 hardcoded strategies

orrc replied:
It's mentioned right there in the TODO. At some point this should be fixed to fetch the documentation from the implementations.
The fact that it's an "extension point" is irrelevant to users, unless we have an explicit section in the plugin manager for BuildDiscarders that we can point to.
On 17 Mar 2016 08:14, Oleg Nenashev <notifications@github.com> wrote:In core/src/main/resources/jenkins/model/BuildDiscarderProperty/help.html:
> @@ -1,19 +1,51 @@
>  <div>
> -  This controls the disk consumption of Jenkins by managing how long you'd like to keep
> -  records of the builds (such as console output, build artifacts, and so on.)
> -  Jenkins offers two criteria:
> -
> +  This determines when, if ever, build records for this project should be
> +  discarded. Build records include the console output, archived artifacts, and
> +  any other metadata related to a particular build.
> +  <p>
> +  Keeping fewer builds means less disk space will be used in the <i>Build Record
> +  Root Directory</i>, which is specified on the <i>Configure System</i> screen.
> +  <p>
> +  <!-- TODO: Most/all of the following should be dynamic, showing info from the
> +       `BuildDiscarder` implementations that are currently loaded -->
> +  Jenkins offers two options for determining when builds should be discarded:

This is an extension point, needs to be mentioned. Currently it smells like 2 hardcoded strategies

—You are receiving this because you authored the thread.Reply to this email directly or view it on GitHub

recena commented on core/src/main/resources/jenkins/model/BuildDiscarderProperty/help.html:
```diff
@@ -1,19 +1,51 @@
+  You can also ensure that important builds are kept forever, regardless of the
+  setting here &mdash; click the <i>Keep this build forever</i> button on the
+  build page.
+  <br>
```
I recommend to use `<p>` instead of `<br>`.

orrc replied:
It would be good to know _why_ you want to recommend this :)

I kept the two lines together, since they are related.

recena replied:
We use the different HTML tags according to their representation and it is an mistake. if you want to define blocks of text, use always `<p>`.

From W3C, it is clear its usage: https://www.w3.org/wiki/HTML/Elements/br

orrc replied:
> The &lt;br&gt; element represents a line break.

I'm using it as a line break here.  As mentioned, I don't want to define a block of separate text in this case.

These are standalone snippets of HTML, and not part of a larger semantically structured document, so I'm not too bothered about having markup that will make Tim BL weep tears of joy :)

recena replied:
> these are standalone snippets of HTML, and not part of a larger semantically structured document

Well, this kind of snippets will part of a HTML document.

batmat replied:
The thing is, if we kind of define and follow some semantic things like those, we could very probably even if necessary define some specific classes to be used inside those HTML doc snippets, and make them visually consistent.

**(NOTE: thanks very much anyway & again, in case it's not obvious. What you did there is in any case a gigantic improvement of what was previously present).**

recena commented on core/src/main/resources/jenkins/model/BuildDiscarderProperty/help.html:
```diff
@@ -1,19 +1,51 @@
+  <hr>
+
+  In the <i>Advanced</i> section, the same options can be specified, but
+  specifically for build <b>artifacts</b>.  If enabled, build artifacts will be
```
Please, `<strong>` instead of `<b>`.

orrc replied:
Yeah, I wondered about that at the time, and whether `<em>` for italics.

recena replied:
From W3C is clear: https://www.w3.org/International/questions/qa-b-and-i-tags

batmat replied:
Commented above for the ~ same things, should have scrolled down before maybe. :)

recena commented on core/src/main/resources/jenkins/model/Messages.properties:
```diff
@@ -70,4 +70,4 @@ ParameterizedJobMixIn.build_with_parameters=Build with Parameters
 BlockedBecauseOfBuildInProgress.shortDescription=Build #{0} is already in progress{1}
 BlockedBecauseOfBuildInProgress.ETA=\ (ETA:{0})
-BuildDiscarderProperty.displayName=Discard Old Builds
+BuildDiscarderProperty.displayName=Discard old builds
```
If it changes this, we should other uppercases examples: `Build now`. :+1: for this change.

orrc replied:
Most config options are in lowercase, since some can be quite long, but for actions in the sidebar like "Build Now" can probably remain as they are.

But yes, in general, I plan to make these consistent.

recena replied:
@orrc :+1:

recena commented:
LGTM :+1:

recena commented:
@orrc My comments are not important, only tips with my recommendations. This PR LGTM. Thanks.

recena commented:
@orrc What do you think if we review the `styleguide` https://github.com/jenkinsci/jenkins/pull/2101 with my comments?

orrc committed:
Rename "Text Parameter" to "Multi-line String Parameter".

Previously, having both "String Parameter" and "Text Parameter" was completely
confusing for end users.  Both parameters essentially have the same underlying
implementation; just the UI is different: one has a single-line text field,
while the other has a multi-line text field.

(Preferably, both would also be called "text" parameters rather than "string"
parameters.)

orrc committed:
Expand parameterised builds documentation.

orrc commented:
@recena Sure, adding rules to the styleguide would be good.

I'm just not 100% sure what you're concretely suggesting — we should use `<strong>` instead of `<b>`, and perhaps `<em>` instead of `<i>`?

If you can summarise the changes you are suggesting in #2101, that would be great.

orrc committed:
Update the 'disable project' documentation.

orrc committed:
Update the 'concurrent builds' documentation.

orrc commented on core/src/main/resources/hudson/model/AbstractProject/help-concurrentBuild.html:
```diff
@@ -1,24 +1,39 @@
-    workspace directory name, e.g. "@2".  The separator "@" can be configured by setting the system property
-    "hudson.slaves.WorkspaceList" to the desired separator string on the Jenkins command line. 
-    E.g. "-Dhudson.slaves.WorkspaceList=-" will use a dash as separator.
+<div>
```
Test comment 1.

KostyaSha replied:
why you changed intend from 2 space to 4?

daniel-beck commented on core/src/main/resources/hudson/model/Messages.properties:
```diff
@@ -319,9 +319,9 @@ Permalink.LastFailedBuild=Last failed build
 
 ParameterAction.DisplayName=Parameters
-ParametersDefinitionProperty.DisplayName=This build is parameterized
+ParametersDefinitionProperty.DisplayName=This project is parameterized
```
Not sure this shouldn't just say *Builds are parameterized* or *Builds have parameters*, because it's the builds that have the parameters.

daniel-beck commented on core/src/main/resources/jenkins/model/BuildDiscarderProperty/help.html:
```diff
@@ -1,19 +1,51 @@
+  <p>
+  For example, if a project builds some software and produces a large installer,
+  which is archived, you may wish to always keep the console log and information
+  about which source control commit was built.  While for disk space reasons,
```
Wouldn't put that into a separate sentence.

orrc replied:
Yeah, it was a bit awkward. It's back in one long sentence now.

daniel-beck commented on core/src/main/resources/jenkins/model/BuildDiscarderProperty/help.html:
```diff
@@ -1,19 +1,51 @@
\ No newline at end of file
+  For time-based limits, note that Jenkins does not discard items as soon as
+  the configured time is exceeded; these rules are evaluated the next time that
+  a build of this project starts.
```
*All rules* are evaluated only when the next build *finishes*. This is relevant e.g. when you reduce the number in the config and wonder why Jenkins does not start deleting builds.

orrc replied:
Yeah, not sure why I left out the other types of rules.  Thanks for the info; updated.

daniel-beck commented:
@orrc Are you planning to delete the non-English localizations that are no longer used?

batmat commented on core/src/main/resources/hudson/model/AbstractProject/help-concurrentBuild.html:
```diff
@@ -1,24 +1,39 @@
+  When this option is checked, multiple builds of this project may be executed
+  in parallel.
+  <p>
+  By default, only a single build of a project is executed at a time &mdash; any
```
might be only me, but not convinced about using entities. IMO those days it's almost never to be used (apart from things to despecialize like \< and the likes).

batmat commented on core/src/main/resources/hudson/model/AbstractProject/help-concurrentBuild.html:
```diff
@@ -1,24 +1,39 @@
+  completely independent from one another.
+  <p>
+  Each concurrently executed build occurs in its own build workspace, isolated
+  from any other builds.  By default, Jenkins appends "<tt>@&lt;num&gt;</tt>" to
```
\<tt> is actually obsolete. Prefer using \<code> for example (not sure where/if there's an existing CSS in the core for help texts).

batmat commented on core/src/main/resources/hudson/model/AbstractProject/help-concurrentBuild.html:
```diff
@@ -1,24 +1,39 @@
+  href="https://wiki.jenkins-ci.org/display/JENKINS/Features+controlled+by+system+properties"
+  target="_blank">wiki page</a>.
+  <p>
+  However, if you enable the <i>Use custom workspace</i> option, all builds will
```
\<em> instead, preferably. More semantic. [IMO again]

batmat commented on core/src/main/resources/hudson/model/ParametersDefinitionProperty/help.html:
```diff
@@ -1,14 +1,32 @@
+  <p>
+  See the <a href="https://wiki.jenkins-ci.org/display/JENKINS/Parameterized+Build"
+  target="_blank">Parameterized Builds wiki page</a> for more information about
+  this feature.
```
:+1: 

Wouldn't it make sense to explain here the fact that Jenkins will only add a new build in the queue *if* there's no existing queued build with the *same* parameters?

We sometimes have questions about that from newcomers, and this is a interesting thing to know IMO. ?

batmat commented:
:+1: 
Great work. I made some remark, but almost nothing on the content. 
Maybe the only thing, like @daniel-beck, would be to remove the now unused translations from the *.properties files.

daniel-beck commented:
Needs 2.0 merge.

orrc committed:
Merge branch '2.0' into JENKINS-33364_freestyle-config

orrc committed:
Update and expand documentation for label expressions.

orrc committed:
Update documentation for quiet period.

orrc committed:
Update and expand documentation for SCM retries.

daniel-beck commented:
@orrc How finished is this?

orrc committed:
Update translation keys for config-disableBuild.

orrc committed:
Add note to the documentation about how parameterized builds are enqueued.

orrc committed:
Clarify when the "Discard old builds" configuration is evaluated.

orrc committed:
Merge remote-tracking branch 'origin/2.0' into JENKINS-33364_freestyle-config
